CORE-13365 Client Monitoring: rate limit and limit page size based on memory

pgellert opened the pull request:
* [admin: rate limit ListKafkaConnections](https://github.com/redpanda-data/redpanda/pull/28251/commits/f437f59449b80c4f32bddc39f443157af9cb047c)
Rate limit the endpoint to 1 external cluster-wide request per broker.
Note that each cluster-wide request turns into 1 sub-request per broker,
so the rate limit allows up to num_brokers in flight requests on each
broker at the same time.

This rate limiting is to limit the CPU impact of the endpoint, a
separate memory semaphore is introduced in the next commit.

* [resource_mgmt: reserve memory for admin API](https://github.com/redpanda-data/redpanda/pull/28251/commits/9e1808edd8a35daa2ff6c60a1c77b452b1c5e5e2)
Reserve about ~2% (depending on which features are enabled) for the
admin API subsystem.

This is initially going to be used by the ListKafkaConnections endpoint,
but could be shared with other endpoints as well.

* [admin: limit ListKafkaConnections page_size to mem](https://github.com/redpanda-data/redpanda/pull/28251/commits/65c9acf8957d80080f75aa3e8411a978f141b8fc)
This endpoint uses approximately 3KiB x kafka_connection_count amount of
memory, which can get large in clusters with a large connection count.

Cap the page_size based on the available memory sempahores to the admin
API and reserve memory semaphores for it on all cores.

Fixes https://redpandadata.atlassian.net/browse/CORE-13365

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
admin: rate limit ListKafkaConnections

Rate limit the endpoint to 1 external cluster-wide request per broker.
Note that each cluster-wide request turns into 1 sub-request per broker,
so the rate limit allows up to num_brokers in flight requests on each
broker at the same time.

This rate limiting is to limit the CPU impact of the endpoint, a
separate memory semaphore is introduced in the next commit.

pgellert committed:
admin: limit ListKafkaConnections page_size to mem

This endpoint uses approximately 3KiB x kafka_connection_count amount of
memory, which can get large in clusters with a large connection count.

Cap the page_size based on the available memory sempahores to the admin
API and reserve memory semaphores for it on all cores.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/f92fedc6ead10cb8b5a6ab40937d5a27ae8bbbfe..236a7a50c25aaa86357c39a0f36a38596cda8a95): address copilot feedback

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f9fad50f-38a0-43b4-a7f8-5d5137f77f1b-->
<!--TEST RESULTS-->
<!--877061ac-295c-445d-964c-114657645e88-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75203">build#75203</a></summary>


| test_class                    | test_method                                | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                               |
|:------------------------------|:-------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_auto_prefix_trimming                  | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75203#019a3055-445b-47fb-ba0d-70c3d24611f8 | FLAKY         | 19/21    | upstream reliability is '92.89940828402366'. current run reliability is '90.47619047619048'. drift is 2.42322 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming      |
| MountUnmountIcebergTest       | test_simple_remount                        | {"cloud_storage_type": 1}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75203#019a3055-445a-48d2-91e0-daf61169f814 | FLAKY         | 15/21    | upstream reliability is '78.62595419847328'. current run reliability is '71.42857142857143'. drift is 7.19738 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                  |
| DeleteRecordsTest             | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": true, "truncate_point": "at_high_watermark"}       | integration | https://buildkite.com/redpanda/redpanda/builds/75203#019a3055-445d-4b39-b5b9-beaaed1f790e | FLAKY         | 20/21    | upstream reliability is '98.78345498783455'. current run reliability is '95.23809523809523'. drift is 3.54536 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations |
| LogCompactionTxRemovalTest    | test_tx_control_batch_removal              | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75203#019a3053-85a3-4512-9fce-94b0daef46e2 | FLAKY         | 12/21    | upstream reliability is '92.0'. current run reliability is '57.14285714285714'. drift is 34.85714 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal     |
| RapidTopicRecreateTest        | test_topic_rapid_recreation                | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75203#019a3053-85a0-450c-a94e-0749f242faf9 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RapidTopicRecreateTest&test_method=test_topic_rapid_recreation           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75543">build#75543</a></summary>


| test_class        | test_method         | test_arguments                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                        |
|:------------------|:--------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------|
| TieredStorageTest | test_tiered_storage | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75543#019a4f76-4a59-4525-bb30-8ae23dc9ccb4 | FLAKY         | 20/21    | upstream reliability is '99.48717948717949'. current run reliability is '95.23809523809523'. drift is 4.24908 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage |
</details>
<!--/body-->

BenPope commented on src/v/redpanda/admin/kafka_connections_service.cc:
```diff
@@ -186,6 +186,40 @@ ss::future<size_t> gather_all_shards(
+    try {
+        auto units = co_await ss::get_units(*_rate_limiter, 1, timeout);
+        co_return ssx::remote_semaphore_units(
+          std::move(units), rate_limiter_shard);
```
I'd be tempted to avoid `ssx::remote_semaphore` and instead do this:
```suggestion
        co_return ss::make_foreign(
          std::make_unique<ssx::semaphore_units>(std::move(units)));
```

pgellert replied:
Yeah, good idea. I've adopted this now.

BenPope commented on src/v/redpanda/admin/services/cluster.cc:
```diff
@@ -51,11 +81,29 @@ cluster_service_impl::list_kafka_connections(
+    if (
+      new_page_size < original_page_size
+      && resp.get_total_size() > new_page_size) {
+        throw_no_memory(max_connections);
```
I'm not sure I understand the reasoning here.

pgellert replied:
The idea is that:
(a) if we reduced the page_size, we don't want to return a partial response to the caller because that would violate the caller's expectations about the response (if `resp.total_size > req.page_size`, then the caller should expect that the `len(resp.connections) = req.page_size`)
(b) but we also want to fulfil requests with `page_size = INT_MAX`

So here we check if we had to use a restricted page size (`new_page_size < original_page_size`) AND there would have been more responses than our restricted new_page_size (`resp.get_total_size() > new_page_size`). If so, that would result in the scenario described in (a) and so we throw instead of returning an inconsistent response.

BenPope replied:
OK. I think it would be useful to return a subset, but if that's how to API is designed, then that's fine by me.

pgellert replied:
Yeah, fair enough. It looks like the pagination AIP that I am trying to follow here ([AIP-158](https://google.aip.dev/158)) allows for capping the page_size and returning less data than available, so I went with that to simplify here and still return partial data in this case. That's actually also nicer for the debug bundle use case where we may want to just save as much data as we can get with a single request per broker.

pgellert committed:
resource_mgmt: reserve memory for admin API

Reserve about ~2% (depending on which features are enabled) for the
admin API subsystem.

This is initially going to be used by the ListKafkaConnections endpoint,
but could be shared with other endpoints as well.

pgellert commented:
force-push:
* [first](https://github.com/redpanda-data/redpanda/compare/236a7a50c25aaa86357c39a0f36a38596cda8a95..0975494483008979dc01a866acef7dc6e6770f9d): Rework based on review feedback to reserve memory from pre-allocated memory semaphore units reserved for the admin API. The memory reserved for the admin API is ~5%. This is sufficient for listing all the connections of a single broker in current all RP Cloud deployments. Internally the page_size is capped to this available memory limit, and we may return a partial response if an overly large page_size was requested. If clients reach this memory limit, they would have to rely on filtering to request data broker-by-broker, shard-by-shard or by sorting+filtering on uid or using some similar client-side "pagination" technique.
* [second](https://github.com/redpanda-data/redpanda/compare/0975494483008979dc01a866acef7dc6e6770f9d..2f64c35caa85a38d4b772917a5f238d7140ff2d1): rebase to dev to fix merge conflicts

pgellert commented on src/v/redpanda/admin/services/cluster.cc:
```diff
@@ -29,6 +31,34 @@ namespace {
 // NOLINTNEXTLINE(*-non-const-global-variables,cert-err58-*)
 ss::logger brlog{"admin_api_server/cluster_service"};
 
+size_t calculate_max_connections() {
```
I've reworked this a bit now, but let me know if you think this still applies.

pgellert commented on src/v/redpanda/admin/services/cluster.cc:
```diff
@@ -29,6 +31,34 @@ namespace {
+size_t calculate_max_connections() {
+    // Get available memory on the request handling shard. This includes
+    // reclaimable memory (batch cache).
+    const auto available_mem = resources::available_memory::local().available();
```
> Maybe we could assign 1 share or something (from memory_groups) to the admin API, and draw from a semaphore of the right size instead? Like we do in the Kafka API.

As we discussed on slack, I went with this approach, but with 5 shares now. This allows for requesting data per-broker for debug bundles and other list-all-connections use cases.

rockwotj commented on src/v/redpanda/admin/services/cluster.cc:
```diff
@@ -48,6 +48,9 @@ cluster_service_impl::list_kafka_connections(
 
     auto& kcs = _kafka_connections_service.local();
+
+    auto units = co_await kcs.rate_limit();
```
This can deadlock with proxied requests as we are both waiting on one another to complete a request. I guess the semaphore will just timeout, but maybe we need console/rpk to consistently talk to a single node to prevent that.

pgellert replied:
I've fixed the deadlock now by skipping rate limiting for proxied requests. The rate limiter's main purpose is to protect CPU usage and provide some backpressure in case a rogue client starts to spam the broker with requests. Since proxied requests have already been rate limited at the first broker, they do not need further rate limiting to achieve that. The admin cpu scheduling group also already provides some level of backpressure / limiting for these requests, so we're still protected from overloading the broker with too many requests.

I think asking console/rpk to talk to a single node defeats the purpose of admin v2 trying to be target node-agnostic, so I wanted to avoid relying on that. I did look into proxying all cluster-wide requests through the controller leader internally, but that adds complexity to both the endpoint and the proxy client, and it still wouldn’t eliminate deadlocks during leadership changes. Another idea I had was to temporarily release the rate-limiting semaphore during external RPCs, but that quickly becomes intrusive and hard to reason about. Skipping rate limiting for proxied requests seems to me as the simplest and safest fix.

rockwotj commented on src/v/resource_mgmt/memory_groups.cc:
```diff
@@ -42,13 +42,15 @@ struct memory_shares {
     constexpr static size_t rpc = 20;
     constexpr static size_t recovery = 10;
     constexpr static size_t tiered_storage = 10;
+    constexpr static size_t admin = 5;
```
Feels a bit heavy to take 5 shares away for the admin server. Does it really deserve half the same amount of memory that data_transforms or datalake uses?

I wonder if we should just share memory with the kafka server memory semaphore...

pgellert replied:
Fair point. I have now reduced the number of shares from 5 to 2.

Originally, I was aiming for having enough memory to be able to print at least a broker's worth of connections in RP Cloud. However, after talking to @travisdowns some more, I now think it would be simpler to instead do client-side pagination for the list-all-connections use case of the rpk debug bundle, by sorting on `uid` and filtering by `uid > {last_uid}`. This is simple to implement client-side and the overhead is acceptable in performance based on [scale testing](https://github.com/redpanda-data/redpanda/pull/28176). It adds some overhead to this use case, because now we have to sort the result, whereas if we had more memory, we could return unordered results broker-by-broker. However, this client-side pagination approach has the benefit that it works regardless of the (current or future) connection limits of the brokers, and the number of requests scale with the number of connections in the cluster, rather than by the number of brokers. So with this client-side pagination approach in mind, we can reduce the memory usage.

> I wonder if we should just share memory with the kafka server memory semaphore...

Travis told me that we are not pressed for memory, so I would avoid the complex interaction of this and rather reserve memory separately, but reserve less memory. I think we could optimize this in the future if the brokers become memory-bottlenecked.

pgellert replied:
I tested with this command in 2-3 shells and saw that we no longer see retries/timeouts on the rate limiter, and only see timeouts on the memory semaphore if those are exhausted (multiple clients spamming with `--limit {large_number}`)
```
watch -n 0.01 'rpk cluster connections list -X=admin.hosts=0.0.0.0:9644,0.0.0.0:9645,0.0.0.0:9646 --verbose'
```

BenPope commented on src/v/redpanda/admin/kafka_connections_service.h:
```diff
@@ -42,8 +47,13 @@ class kafka_connections_service {
+    ss::future<remote_units> do_rate_limit();
+
     ss::sharded<kafka::server>& _kafka_server;
+    std::unique_ptr<ssx::semaphore> _rate_limiter{};
```
Is it worth having a mechanism to call `broken()` on this, perhaps a `stop()`?

pgellert replied:
Good point. I think it's not worth it for the rate_limiter semaphore here, because it doesn't speed up shutdown if I just add `_rate_limiter.broken()` into a new `kafka_connections_service::stop()`. That's because the admin server stops earlier than the kafka connections service, so there would be no in-flight requests by the time `kafka_connections_service::stop()` is called. We could call `kafka_connections_service::stop()` explicitly earlier than the admin server is stopped, but I think it's not necessary here since the timeout is short, and so I would avoid the complexity.

However, it is simpler and makes sense to add a `broken()` to the admin server's memory semaphore, so I've added that now.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/2f64c35caa85a38d4b772917a5f238d7140ff2d1..6eb6a943a0f29520ba6d0765587a1904b488223a):
- relax rate limit to only apply to non-proxied requests to avoid deadlocks when two requests arrive on different nodes
- add jitter to timeouts to desynchronize requests in case multiple requests time out on the memory or rate limiting semaphores
- reduce memory shares from 5 to 2
- fix bug in the way the max memory semaphore units was calculated (max_count vs initial current) + add regression test
- break memory semaphore on admin server `stop()`

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/6eb6a943a0f29520ba6d0765587a1904b488223a..65c9acf8957d80080f75aa3e8411a978f141b8fc): fix unit test failures by only starting `kafka_connections_service` if the admin API is configured

rockwotj commented on src/v/redpanda/admin/kafka_connections_service.cc:
```diff
@@ -96,6 +102,85 @@ ss::future<size_t> gather_all_shards(
+      ss::this_shard_id() == rate_limiter_shard,
+      "do_rate_limit() must be called on shard {}",
+      rate_limiter_shard);
+    const auto timeout = std::chrono::milliseconds(
+      random_generators::get_int(3000, 5000));
```
nit: normally we use https://github.com/redpanda-data/redpanda/blob/dev/src/v/random/simple_time_jitter.h for this

rockwotj reviewed (approved):
LGTM

BenPope commented on src/v/redpanda/admin/services/cluster.cc:
```diff
@@ -57,6 +59,26 @@ cluster_service_impl::list_kafka_connections(
         rate_units = co_await kcs.rate_limit();
     }
 
+    if (req.get_page_size() < 0) {
+        throw serde::pb::rpc::invalid_argument_exception(
+          "page_size must be non-negative");
+    }
```
This should be moved to just after the license check.

BenPope commented on src/v/redpanda/admin/kafka_connections_service.cc:
```diff
@@ -109,8 +112,17 @@ kafka_connections_service::kafka_connections_service(
+ss::future<>
+kafka_connections_service::start(ssx::semaphore& admin_memory_semaphore) {
+    _admin_memory_semaphore = &admin_memory_semaphore;
+    _admin_memory_semaphore_max = admin_memory_semaphore.current();
```
It's not obvious why `current()` is correct here, can a comment be added to explain that no units are expected to have been taken at this point due to lifetime, so it's a reasonable proxy for the capacity.

BenPope commented on src/v/redpanda/admin/kafka_connections_service.cc:
```diff
@@ -134,6 +146,41 @@ kafka_connections_service::do_rate_limit() {
+
+ss::future<kafka_connections_service::remote_units>
+kafka_connections_service::do_memory_limit(size_t connection_count) {
+    try {
+        vassert(
+          _admin_memory_semaphore != nullptr,
+          "Admin memory semaphore not initialized");
+        const auto timeout = std::chrono::milliseconds(
+          random_generators::get_int(3000, 5000));
+        auto units = co_await ss::get_units(
+          *_admin_memory_semaphore,
+          connection_count * memory_per_connection,
+          timeout);
+        co_return ss::make_foreign(
+          std::make_unique<ssx::semaphore_units>(std::move(units)));
+    } catch (const ss::semaphore_timed_out&) {
+        throw serde::pb::rpc::resource_exhausted_exception(
+          fmt::format(
+            "Timeout waiting for memory for list_kafka_connections. Consider "
+            "filtering for a single broker or reducing page_size."));
+    }
+}
```
nit: This might be cleaner with continuation-style.

BenPope commented on tests/rptest/tests/list_kafka_connections_test.py:
```diff
@@ -73,8 +73,9 @@ def test_list_kafka_connections(self):
             self.redpanda,
             auth=(self.superuser.username, self.superuser.password),
         )
+        int32_max = 2_147_483_647
```
nit:
```suggestion
        int32_max = (2**31) - 1
```

BenPope reviewed (approved):
LGTM, just a couple of minor things.

pgellert commented:
I'll keep the nits/recommendations in mind for if I open another PR in this area but will merge this in now to get this into the next RC.
